docs/proposals/disk-affinity.md

```diff
@@ -0,0 +1,161 @@
+# Proposal: Supporting disk affinity/anti-affinity feature 
+
+Last updated: 18.01.2022
+
+## Abstract
+
+This proposal contains approach for supporting disk affinity/anti-affinity feature.
+
+## Background
+
+Current lvg storageclass cannot mount physical disks to PVs with the awareness of the requesting pods. So, it's possible
+for different applications to share the same disk.
+CSI should provide a way to support:
+* mount PVs for the same pod on the same disk (disk affinity);
+* do not mount PVs on the disks that already have some PVs from other specific pods (disk antiaffinity).
+
+## Proposal
+
+Usually for affinity/antiaffinity topology feature is used. But the minimal domain, on which this feature operates, is nodes, not disks.
+So we should use a different approach for this one. Another approach, which is proposed, is to use annotations for pointing out how to use disk affinity/antiaffinity:
+affinity.drives.csi-baremetal.dell.com/type: <pod-bound> - mount PVs for the same pod on the same disk;
```

Comment:
Suggest to have key just `affinity.csi-baremetal.dell.com` without drives

Reply:
- What if we want to have 2 PVs from a single pod to be placed on one disk and 1 on another?
- Are those rules are strict? Can we provide some API for user to not prevent scheduling when condition is not meet?
- Is `affinity.drives.csi-baremetal.dell.com/type` have only single key?

Reply:
For the question of affinity naming - In future we may have some other affinity features, not related with drives. And it may become the problem if we exclude drive from naming here.

Reply:
There wasn't such use case, but for this we can introduce new affinity type: 
affinity.drives.csi-baremetal.dell.com/type: volume-bound
Then we can have another annotation for affinity specifying which pvs should be placed on the same disks:
affinity.drives.csi-baremetal.dell.com/volumes: "[pv-1, pv-2], [pv-3, pv4]"

Likewise we can do the same for antiaffinity:
antiaffinity.drives.csi-baremetal.dell.com/type: volume-bound
And specify which drives should not be placed together on the same drives:
antiaffinity.drives.csi-baremetal.dell.com/volumes: "[pv-1, pv-2], [pv-1, pv3]"

Reply:
For now affinity.drives.csi-baremetal.dell.com/type has single type. But in future it may have more types, as we've already discussed. So we may rename it to affinity.drives.csi-baremetal.dell.com/types

Reply:
> For the question of affinity naming - In future we may have some other affinity features, not related with drives. And it may become the problem if we exclude drive from naming here.

- LVG doesn't match to the single drive in common case since one LVG can be spread across multiple drives. Thus `drives` sounds confusing here. May use `volumes` instead if you think that just `affinity` is to common.

Reply:
> There wasn't such use case, but for this we can introduce new affinity type: affinity.drives.csi-baremetal.dell.com/type: volume-bound Then we can have another annotation for affinity specifying which pvs should be placed on the same disks: affinity.drives.csi-baremetal.dell.com/volumes: "[pv-1, pv-2], [pv-3, pv4]"
> 
> Likewise we can do the same for antiaffinity: antiaffinity.drives.csi-baremetal.dell.com/type: volume-bound And specify which drives should not be placed together on the same drives: antiaffinity.drives.csi-baremetal.dell.com/volumes: "[pv-1, pv-2], [pv-1, pv3]"

please add this to proposal

Reply:
> For required/preferred behaviour there are two possible solutions: We have modify our current annotations, e.g., in the following way: required.antiaffinity.drives.csi-baremetal.dell.com/labels: ecs-cluster-ss, ecs-cluster-pvg preferred.antiaffinity.drives.csi-baremetal.dell.com/labels: ecs-cluster-ss, ecs-cluster-pvg
> 
> Or we can modify type of affinity: antiaffinity.drives.csi-baremetal.dell.com/type: pod-label-required
> 
> The second solution may be more elegant

please add this to proposal

Reply:
Added new types and options to proposal